Handle situation when availibity api is declined

dolohow opened the pull request:
I think when you get the ban from Ryanair from hammering their API you'll receive message
```
{'code': 'SHIELD_BLOCK', 'message': 'Availability declined'}
```

Let's handle that better, so user knows what to do.

dolohow committed:
Handle situation when availibity api is declined

I think when you get the ban from Ryanair from hammering their API
you'll receive message
```
{'code': 'SHIELD_BLOCK', 'message': 'Availability declined'}
```

Recently, they changed it to just message
```
{'message': 'Availability declined'}
```

Let's handle that better, so user knows what to do.

cohaolain commented:
Very nice. These new exceptions are a good idea. 

I've never actually hit this particular rate limit - what's the current behaviour when this occurs (can't check currently, I'm just off a flight 😊)?
Do you happen to know if this shield is IP-based, or something else? Perma-ban? Can you still use the booking flow on the website, logged in or not? Curious what kind of load you were generating when you encounter this - perhaps a new warning would be useful when getting somewhat close, either in docs or library.

Thanks, sorry for so many questions!

dolohow commented:
> I've never actually hit this particular rate limit - what's the current behaviour when this occurs (can't check currently, I'm just off a flight blush)?

Not sure if I understand correctly, but basically my code is just skipping and printing an error message about the problem.  It is not exiting, because the next call to `get_all_flights` might be successful.

> Do you happen to know if this shield is IP-based, or something else? Perma-ban? Can you still use the booking flow on the website, logged in or not? Curious what kind of load you were generating when you encounter this - perhaps a new warning would be useful when getting somewhat close, either in docs or library.

Probably it's IP-based, because on different network I am not having this issue.  I am calling Ryanair API every 10 minutes, now I tried to raise to 30 minutes, but it did not help.  Also, not every call is rejected, only some and sometimes.  For example, today, sometimes all and sometimes only two calls to `get_all_flights` caused difficulties.  I added print every time when I call `get_all_flights` with patch enabled.
```
2023-04-22 05:13:43.462 ERROR:Ryanair API: Availibility declined, you are most likely banned
get_all_flights
2023-04-22 05:13:44.963 ERROR:Ryanair API: Availibility declined, you are most likely banned
get_all_flights
2023-04-22 05:13:45.556 ERROR:Ryanair API: Availibility declined, you are most likely banned
get_all_flights
get_all_flights
2023-04-22 05:13:49.025 ERROR:Ryanair API: Availibility declined, you are most likely banned
get_all_flights
get_all_flights
2023-04-22 05:13:51.967 ERROR:Ryanair API: Availibility declined, you are most likely banned
get_all_flights
2023-04-22 05:13:52.488 ERROR:Ryanair API: Availibility declined, you are most likely banned
get_all_flights
2023-04-22 05:13:55.413 ERROR:Ryanair API: Availibility declined, you are most likely banned
get_all_flights
```

EDIT:
I actually miscalculated this
I am calling `api_get_all_flights` 54 times each 10/15 minutes in period of 1/2 seconds

cohaolain commented:
So just to confirm, after you get the `{'code': 'SHIELD_BLOCK', 'message': 'Availability declined'}` message, without changing anything, later queries from the same IP address might start working again?

To confirm, you've seen this happen with just *one* API every 30 minutes?

If that's the case, and it's not a permanent "ban", we should probably call it something else - though from what you've said, I'm not sure this *quite* matches the behaviour of a standard API rate-limiting implementation either.

dolohow commented:
> So just to confirm, after you get the {'code': 'SHIELD_BLOCK', 'message': 'Availability declined'} message, without changing anything, later queries from the same IP address might start working again?

Yes

> To confirm, you've seen this happen with just one API every 30 minutes?

Every 10 minutes, also 30 minutes as well.  Perhaps my IP address is somehow marked, I have no idea what cool downs mechanism/strategies they are using.

> If that's the case, and it's not a permanent "ban", we should probably call it something else - though from what you've said, I'm not sure this quite matches the behaviour of a standard API rate-limiting implementation either.

You can rework some of those wording.  Maybe throttled or something...

dolohow commented:
@cohaolain  They changed response a little bit, so I tweaked it.  Also changed some wording

dolohow commented:
I think it's blocked all over
```
curl "https://www.ryanair.com:443/api/booking/v4/en-ie/availability?ADT=1&TEEN=0&CHD=0&INF=0&DateOut=2023-09-09&DateIn=&Origin=PMO&Destination=KRK&OriginIsMac=False&DestinationIsMac=False&IncludeConnectingFlights=False&ToUs=AGREED"
```

however it does work in browser, so perhaps spoofing headers should work

EDIT:
Spoofing is not enough, cookie is now required...

cohaolain commented:
> I think it's blocked all over
>  
> EDIT: Spoofing is not enough, cookie is now required...

Interesting, that's disappointing :disappointed:

cohaolain commented:
Have also just tried it myself, and it looks like the only _required_ header is the cookie.

dolohow commented:
Are you going to merge this?

cohaolain commented on ryanair/ryanair.py:
```diff
@@ -25,6 +25,16 @@
+
+class ShieldException(RyanairException):
+    def __init__(self):
+        super().__init__("Availibility declined, perhaps you are being throttled")
```
Could we please update this to simply "Availablity API declined to provide a result", as it looks like this is going to be essentially normal for now, not caused by throttling.

I think it should also no longer be called `ShieldException`, but perhaps more simply `AvailabilityException` or `AvailabilityDecli edException`.

I'd also consider including the `response.json()` in the logged error, in case this includes more info in the future.

dolohow replied:
I did modifications, however, calling response.json() somewhere here is not a good option IMHO.  Also, we catch error when there is a particular error message, so that would be dubious

cohaolain commented:
> Are you going to merge this?

Yes, subject to comments above, thanks for your time 👍

dolohow committed:
Use session to acquire cookies during API requests

Ryanair recently introduced some protections in their APIs.  Without
this API now always throws "Availibility declined".  However, I
discovered that acquiring cookies during the first request allows for
API requests to work as expected.

dolohow commented:
Also now, after I modified code it's working fine. However, they still blocking on IP address, because after some time it stopped working and I had to change ip address.

cohaolain commented on ryanair/ryanair.py:
```diff
@@ -191,7 +206,9 @@ def _on_query_error(e):
 
-        return requests.get(url, params=params).json()
+        # Visit main website to get session cookies
+        self.session.get('https://www.ryanair.com/ie/en')
```
Excellent, though I would much prefer to only perform this particular `GET` once - either in the constructor, or to cache it on the first call to ` _retryable_query`. My understanding is that this would result in two `GET` requests for every single API call, which seems like too much overhead.

dolohow replied:
Yes, but cookie is expiring after some time (tested) but I don't know for how long. And you need to implement retry solution. I don't have a need to do that. You can develop better solution later on, but for now at least it's working.

cohaolain commented:
> Also now, after I modified code it's working fine. However, they still blocking on IP address, because after some time it stopped working and I had to change ip address.

Thanks for your work on this. I think this probably warrants, at least, a disclaimer in the README that it's heavily rate-limited at least. I'll write this myself once this PR is merged.
